crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,742 @@
+                    self.current_blank_characters = 0;
+                    return Some(logical_line);
+                }
+                _ => {}
+            }
+        }
+
+        None
+    }
+}
+
+impl BlankLinesChecker {
+    /// E301, E302, E303, E304, E305, E306
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn check_content(
+        &mut self,
+        tokens: &[LexResult],
+        locator: &Locator,
+        stylist: &Stylist,
+        diagnostics: &mut Vec<Diagnostic>,
+    ) {
+        let mut prev_indent_level: Option<usize> = None;
+        let line_preprocessor = LinePreprocessor::new(tokens, locator);
+
+        for logical_line in line_preprocessor {
+            self.check_line(
+                &logical_line,
+                prev_indent_level,
+                locator,
+                stylist,
+                diagnostics,
+            );
+            if !logical_line.is_comment_only {
+                prev_indent_level = Some(logical_line.indent_level);
+            }
+        }
+    }
+
+    #[allow(clippy::nonminimal_bool)]
+    fn check_line(
+        &mut self,
+        line: &LogicalLineInfo,
+        prev_indent_level: Option<usize>,
+        locator: &Locator,
+        stylist: &Stylist,
+        diagnostics: &mut Vec<Diagnostic>,
+    ) {
+        let indent_size: usize = 4;
+
+        if let Status::Inside(nesting_indent) = self.class_status {
+            if line.indent_level < nesting_indent {
+                if line.is_comment_only {
+                    self.class_status = Status::CommentAfter(nesting_indent);
+                } else {
+                    self.class_status = Status::Outside;
+                }
+            }
+        }
+
+        if let Status::Inside(nesting_indent) = self.fn_status {
+            if line.indent_level < nesting_indent {
+                if line.is_comment_only {
+                    self.fn_status = Status::CommentAfter(nesting_indent);
+                } else {
+                    self.fn_status = Status::Outside;
+                }
+            }
+        }
+
+        // A comment can be de-indented while still being in a class/function, in that case
+        // we need to revert the variables.
+        if !line.is_comment_only {
+            if let Status::CommentAfter(indent) = self.fn_status {
+                if line.indent_level >= indent {
+                    self.fn_status = Status::Inside(indent);
+                }
+                self.fn_status = Status::Outside;
+            }
+
+            if let Status::CommentAfter(indent) = self.class_status {
+                if line.indent_level >= indent {
+                    self.class_status = Status::Inside(indent);
+                }
+                self.class_status = Status::Outside;
+            }
+        }
+
+        // Don't expect blank lines before the first non comment line.
+        if self.is_not_first_logical_line {
+            if line.preceding_blank_lines == 0
```

Comment:
Nit: 

```suggestion
                && line.first_token == Tok::Def
```

Reply:
I assumed you meant that the visual indent was wrong and fixed it in 3941e8f99.